### PR TITLE
Migrating catalog and subquery tests from GPDB-master-TINC-BFV

### PR DESCRIPTION
**Pull request by vraghavan78**
There are several test suites that are in TINC and cdbfast that are not visible to the OSS community.
This lack of visibility causes a lot of pain when the tests fails and we have to chase the root cause.

To avoid, this the query optimizer team is planning to:

* Clean up these test suites
* Move these tests to pg_regress (particularly installcheck-good)

Once this initial effort is done, if needed we can do another pass to optimize these tests.

### CODE REVIEW

File: src/test/regress/sql/catalog.sql [hunk 1]
```diff
@@ -0,0 +1,210 @@
+set optimizer_enable_master_only_queries = on;
```
Comment (xinzweb) [inline src/test/regress/sql/catalog.sql]: rename file to 'bfv_catalog.sql'?

File: src/test/regress/sql/catalog.sql [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+                                c.contype,
+                                c.condef,
+                                i.tableparts
+                ) as ptable_user_con_info(tableid, tableconname, contype, condef, numexpected, numactual, numnames)
+        where
+            contype != 'f' and -- no FK consistency in Rio
+            numexpected != numactual;
+
+--
+--
```
Comment (xinzweb) [inline src/test/regress/sql/catalog.sql]: Is there any description of intent of this test? E.g. testing resource queue etc.?

File: src/test/regress/expected/bfv_subquery.out [hunk 3]
```diff
@@ -0,0 +1,230 @@
+create language plpythonu;
+create or replace function count_operator(explain_query text, operator text) returns int as
```
Comment (xinzweb) [inline src/test/regress/expected/bfv_subquery.out]: Maybe add a description of this function: count number of certain operators in a given plan?

File: src/test/regress/expected/bfv_subquery.out [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+DROP TABLE IF EXISTS t1 CASCADE;
+NOTICE:  table "t1" does not exist, skipping
+DROP TABLE IF EXISTS t2 CASCADE;
+NOTICE:  table "t2" does not exist, skipping
+DROP TABLE IF EXISTS x CASCADE;
+NOTICE:  table "x" does not exist, skipping
+CREATE TABLE t1 (a int, b int) DISTRIBUTED BY (a);
+CREATE TABLE t2 (a int, b int) DISTRIBUTED BY (a);
+CREATE TABLE x (a int) DISTRIBUTED BY (a);
+select count_operator('explain select * from x where a=  (select sum(t1.a)  from t1 inner join (select x.a as outer_ref, * from t2) as foo on (foo.a=t1.a+ outer_ref)  group by foo.a);', 'Table Scan') > 0;
```
Comment (xinzweb) [inline src/test/regress/expected/bfv_subquery.out]: So the intent is to: expect to have 'table scan' operator in query plan? Maybe also state it as a comment, so it's easier to understand.

**Comment by xinzweb** [comment]
Thanks for bring it in. Just minor comments. LGTM.

File: src/test/regress/sql/catalog.sql (hunk 1 above)
Comment (vraghavan78) [inline src/test/regress/sql/catalog.sql]: will do

File: src/test/regress/expected/bfv_subquery.out (hunk 3 above)
Comment (vraghavan78) [inline src/test/regress/expected/bfv_subquery.out]: will do

File: src/test/regress/expected/bfv_subquery.out (hunk 4 above)
Comment (vraghavan78) [inline src/test/regress/expected/bfv_subquery.out]: will do

File: src/test/regress/sql/bfv_subquery.sql [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+    region_nr smallint,
+    region_bezei character varying(50),
+    sm_region_nr smallint,
+    sm_region_bezei character varying(50),
+    land_ges_nr smallint NOT NULL,
+    land_ges_bezei character varying(50) NOT NULL
+) DISTRIBUTED RANDOMLY;
+
+
+COPY test_r_rvv_stada_dim_konst (dim_bezei, dim_elem_lvl, dim_elem_nr, dim_elem_id, dim_elem_bezei, dim_parent_nr, dim_bed) FROM stdin;
```
Comment (asimrp) [inline src/test/regress/sql/bfv_subquery.sql]: Do we need 200+ tuples here?  gpdiff, in particular will be happy if we can reduce the number of tuples.

Comment (vraghavan78) [inline src/test/regress/sql/bfv_subquery.sql]: I will try. But I make no promises, we have more than 10000+ tests we are migrating. This test  take less than a second so not much bang for the buck. But I will try.

**Comment by vraghavan78** [comment]
Thanks all!

**Comment by kuien** [comment]
One observation from GPDB-master-opt-l1 results is that **catalog.sql** fails occasionally.

1. it occurs 1 week 2 days ago 
http://pulse-releng.greenplum.com/browse/projects/GPDB-master-opt-l1/builds/68734/tests/

```diff
*** ./expected/catalog.out	2016-03-24 01:35:40.000000000 -0700
--- ./results/catalog.out	2016-03-24 01:35:40.000000000 -0700
***************
*** 175,184 ****
          where
              contype != 'f' and -- no FK consistency in Rio
              numexpected != numactual;
!  tableid | tableconname | contype | condef | numexpected | numactual 
! ---------+--------------+---------+--------+-------------+-----------
! (0 rows)
! 
  --
  --
  --
--- 175,181 ----
          where
              contype != 'f' and -- no FK consistency in Rio
              numexpected != numactual;
! ERROR:  cache lookup failed for constraint 254596 
  --
  --
  --

======================================================================
```

2. it occurs 2 hours 20 mins ago
http://pulse-releng.greenplum.com/browse/projects/GPDB-master-opt-l1_haozhou/builds/69151/tests/

```diff
*** ./expected/catalog.out	2016-04-02 00:38:30.000000000 -0700
--- ./results/catalog.out	2016-04-02 00:38:30.000000000 -0700
***************
*** 175,184 ****
          where
              contype != 'f' and -- no FK consistency in Rio
              numexpected != numactual;
!  tableid | tableconname | contype | condef | numexpected | numactual 
! ---------+--------------+---------+--------+-------------+-----------
! (0 rows)
! 
  --
  --
  --
--- 175,181 ----
          where
              contype != 'f' and -- no FK consistency in Rio
              numexpected != numactual;
! ERROR:  cache lookup failed for constraint 273518 
  --
  --
  --

======================================================================
```

Any guys meet this issue also? any suggestion? thanks.

**Pull request closed without merging**